### PR TITLE
Added Magento 2 module project template

### PR DESCRIPTION
**Pull request by VitaliyBoyko**
### Description (*)
Added project template for Magento 2 Module

### Fixed Issues (if relevant)
1. magento/magento2-phpstorm-plugin#53: Project Template. Add project template for Magento Module

#### Startup
image

#### Code generation
image

#### Include path suggestion
image

#### Include path confirmation
image

#### Result
image

Also adjusted configs and fixed bugs with action scopes.

### CODE REVIEW

File: src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleRegistrationPhpGenerator.java [hunk 1]
```diff
@@ -28,8 +28,12 @@ public ModuleRegistrationPhpGenerator(@NotNull ModuleRegistrationPhpData moduleR
     }
 
     public PsiFile generate(String actionName) {
-        ModuleDirectoriesData moduleDirectoriesData = directoryGenerator.createOrFindModuleDirectories(moduleRegistrationPhpData.getPackageName(), moduleRegistrationPhpData.getModuleName(), moduleRegistrationPhpData.getBaseDir());
-        return fileFromTemplateGenerator.generate(RegistrationPhp.getInstance(), getAttributes(), moduleDirectoriesData.getModuleDirectory(), actionName);
+        if (moduleRegistrationPhpData.getCreateModuleDirs()) {
+            ModuleDirectoriesData moduleDirectoriesData = directoryGenerator.createOrFindModuleDirectories(moduleRegistrationPhpData.getPackageName(), moduleRegistrationPhpData.getModuleName(), moduleRegistrationPhpData.getBaseDir());
+            return fileFromTemplateGenerator.generate(RegistrationPhp.getInstance(), getAttributes(), moduleDirectoriesData.getModuleDirectory(), actionName);
+        } else {
+            return fileFromTemplateGenerator.generate(RegistrationPhp.getInstance(), getAttributes(), moduleRegistrationPhpData.getBaseDir(), actionName);
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleRegistrationPhpGenerator.java:35]: Same here, we don't need the `else` ramification.

File: src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleXmlGenerator.java [hunk 2]
```diff
@@ -31,8 +31,13 @@ public ModuleXmlGenerator(@NotNull ModuleXmlData moduleXmlData, Project project)
...
     @Override
     public PsiFile generate(String actionName) {
-        ModuleDirectoriesData moduleDirectoriesData = directoryGenerator.createOrFindModuleDirectories(moduleXmlData.getPackageName(), moduleXmlData.getModuleName(), moduleXmlData.getBaseDir());
-        return fileFromTemplateGenerator.generate(ModuleXml.getInstance(), getAttributes(), moduleDirectoriesData.getModuleEtcDirectory(), actionName);
+        if (moduleXmlData.getCreateModuleDirs()) {
+            ModuleDirectoriesData moduleDirectoriesData = directoryGenerator.createOrFindModuleDirectories(moduleXmlData.getPackageName(), moduleXmlData.getModuleName(), moduleXmlData.getBaseDir());
+            return fileFromTemplateGenerator.generate(ModuleXml.getInstance(), getAttributes(), moduleDirectoriesData.getModuleEtcDirectory(), actionName);
+        } else {
+            PsiDirectory etcDirectory = directoryGenerator.findOrCreateSubdirectory(moduleXmlData.getBaseDir(), Package.MODULE_BASE_AREA_DIR);
+            return fileFromTemplateGenerator.generate(ModuleXml.getInstance(), getAttributes(), etcDirectory, actionName);
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleXmlGenerator.java:39]: Same here regarding the `else`.

File: src/com/magento/idea/magento2plugin/generation/MagentoProjectGeneratorSettings.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+package com.magento.idea.magento2plugin.generation;
+
+import com.magento.idea.magento2plugin.project.Settings;
+
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/generation/MagentoProjectGeneratorSettings.java:8]: Remove this empty line

File: src/com/magento/idea/magento2plugin/init/ConfigurationManager.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+            return new DumbAwareAction("Fix") {
+                public void actionPerformed(@NotNull AnActionEvent e) {
+                    notification.expire();
+                    ModuleRootModificationUtil.updateModel(module, (model) -> {
+                        ContentEntry[] contentEntries = model.getContentEntries();
+                        ContentEntry rootEntry = null;
+                        ContentEntry[] entries = contentEntries;
+                        int length = contentEntries.length;
+
+                        for(int i = 0; i < length; ++i) {
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/init/ConfigurationManager.java:90]:
```suggestion
                        for (int i = 0; i < length; ++i) {
```

File: src/com/magento/idea/magento2plugin/init/ConfigurationManager.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        };
+        ApplicationManager.getApplication().invokeLater(runnable, ModalityState.NON_MODAL);
+    }
+
+    public static void notifyGlobally(@Nullable Project project, String title, String message, NotificationType notificationType, Function<Notification, AnAction>... actions) {
+        Notification notification = new Notification("Magento 2 Support", title, message, notificationType);
+        Function[] functions = actions;
+        int length = actions.length;
+
+        for(int i = 0; i < length; ++i) {
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/init/ConfigurationManager.java:240]:
```suggestion
        for (int i = 0; i < length; ++i) {
```

File: src/com/magento/idea/magento2plugin/project/ConfigurableProvider.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Nls
+    @NotNull
+    public String getName() {
+        return "Magento 2";
+    }
+
+    @NotNull
+    public PhpFrameworkConfigurable createConfigurable(@NotNull Project project) {
+
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/project/ConfigurableProvider.java:25]: Remove this empty line

File: src/com/magento/idea/magento2plugin/project/UsagesProvider.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    public UsagesProvider() {
+    }
+
+    @NotNull
+    public String getName() {
+        return "Magento 2";
+    }
+
+    public boolean isEnabled(@NotNull Project project) {
+
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/project/UsagesProvider.java:21]: Empty line

File: src/com/magento/idea/magento2plugin/util/magento/MagentoBasePathUtil.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    public static boolean isMagentoFolderValid(String path) {
+        if (StringUtil.isEmptyOrSpaces(path)) {
+            return false;
+        } else {
+            VirtualFile file = LocalFileSystem.getInstance().findFileByPath(path);
+            if (file != null && file.isDirectory()) {
+                return VfsUtil.findRelativeFile(file, Package.APP, Package.MODULE_BASE_AREA_DIR, ModuleDiXml.FILE_NAME) != null &&
+                        VfsUtil.findRelativeFile(file, new String[]{Package.VENDOR}) != null;
+            } else {
+                return false;
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/util/magento/MagentoBasePathUtil.java:25]: I believe that you may extract this return outside the else, and remove the `else` ramification.

File: src/com/magento/idea/magento2plugin/init/ConfigurationManager.java [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+            return file != null && file.isDirectory() ? file : null;
+        }
+    }
+
+    private static boolean isFileInsideModule(@NotNull VirtualFile magentoFile, @NotNull Module module) {
+        VirtualFile[] contentRoots = ModuleRootManager.getInstance(module).getContentRoots();
+        VirtualFile[] roots = contentRoots;
+        int length = contentRoots.length;
+
+        for(int i = 0; i < length; ++i) {
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/init/ConfigurationManager.java:218]:
```suggestion
        for (int i = 0; i < length; ++i) {
```

File: src/com/magento/idea/magento2plugin/generation/MagentoModuleGenerator.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+import com.magento.idea.magento2plugin.actions.generation.data.ModuleXmlData;
+import com.magento.idea.magento2plugin.actions.generation.generator.ModuleComposerJsonGenerator;
+import com.magento.idea.magento2plugin.actions.generation.generator.ModuleRegistrationPhpGenerator;
+import com.magento.idea.magento2plugin.actions.generation.generator.ModuleXmlGenerator;
+import com.magento.idea.magento2plugin.init.ConfigurationManager;
+import com.magento.idea.magento2plugin.project.Settings;
+import org.jetbrains.annotations.Nls;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
```
[the remark this answers is not in the document]
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/generation/MagentoModuleGenerator.java:31]: ✅

File: src/com/magento/idea/magento2plugin/actions/generation/data/ModuleXmlData.java [hunk 11]
```diff
@@ -32,4 +35,6 @@ public String getModuleName() {
     public PsiDirectory getBaseDir() {
         return this.baseDir;
     }
+
+    public boolean getCreateModuleDirs() { return this.createModuleDirs; }
```
[the remark this answers is not in the document]
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/actions/generation/data/ModuleXmlData.java:39]: ✅

File: src/com/magento/idea/magento2plugin/actions/generation/data/ModuleRegistrationPhpData.java [hunk 12]
```diff
@@ -32,4 +35,6 @@ public String getModuleName() {
     public PsiDirectory getBaseDir() {
         return this.baseDir;
     }
+
+    public boolean getCreateModuleDirs() { return this.createModuleDirs; }
```
[the remark this answers is not in the document]
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/actions/generation/data/ModuleRegistrationPhpData.java:39]: ✅

File: src/com/magento/idea/magento2plugin/actions/generation/data/ModuleComposerJsonData.java [hunk 13]
```diff
@@ -68,4 +71,7 @@ public String getModuleVersion() {
     public List<String> getModuleDependencies() {
         return moduleDependencies;
     }
+
+    public boolean getCreateModuleDirs() { return this.createModuleDirs; }
```
[the remark this answers is not in the document]
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/actions/generation/data/ModuleComposerJsonData.java:75]: ✅

File: src/com/magento/idea/magento2plugin/actions/generation/data/ModuleComposerJsonData.java [hunk 14]
```diff
@@ -68,4 +71,7 @@ public String getModuleVersion() {
...
+
```
[the remark this answers is not in the document]
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/actions/generation/data/ModuleComposerJsonData.java:76]: ✅

File: src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleRegistrationPhpGenerator.java (hunk 1 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleRegistrationPhpGenerator.java:35]: ✅

File: src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleXmlGenerator.java (hunk 2 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleXmlGenerator.java:39]: ✅

File: src/com/magento/idea/magento2plugin/generation/MagentoProjectGeneratorSettings.java (hunk 3 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/generation/MagentoProjectGeneratorSettings.java:8]: ✅

File: src/com/magento/idea/magento2plugin/init/ConfigurationManager.java (hunk 4 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/init/ConfigurationManager.java]: @eduard13 
Thank you for the review. Could you please recheck?

File: src/com/magento/idea/magento2plugin/project/ConfigurableProvider.java (hunk 6 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/project/ConfigurableProvider.java:25]: ✅

File: src/com/magento/idea/magento2plugin/init/ConfigurationManager.java (hunk 5 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/init/ConfigurationManager.java]: ✅

File: src/com/magento/idea/magento2plugin/project/UsagesProvider.java (hunk 7 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/project/UsagesProvider.java:21]: ✅

File: src/com/magento/idea/magento2plugin/util/magento/MagentoBasePathUtil.java (hunk 8 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/util/magento/MagentoBasePathUtil.java:25]: ✅

File: src/com/magento/idea/magento2plugin/init/ConfigurationManager.java (hunk 9 above)
Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/init/ConfigurationManager.java]: ✅

**Pull request merged**